bip-ccv.mediawiki

```diff
@@ -0,0 +1,420 @@
+
+-----
+
+::[[File:bip-ccv/amount_example_4.png|framed|center|alt=split and aggregate amount logic|600px]]
+::'''Figure 4:''' Similar to the previous example, but a second input <code>B</code> also checks the same output <code>X</code> with the <i>default</i> semantic, aggregating its input with the residual amount of the first input.
+
+-----
+
+Note that the ''deduct'' semantic does not allow to check the exact amount of its output. Therefore, in contracts using a scheme similar to figure 3 or 4 above, amounts be constrained with a signature, or with future introspection opcodes that allow fixing the amount.
+
+== Specification ==
+
+The tapscript opcode <code>OP_SUCCESS187</code> (<code>0xbb</code>) is constrained with new rules to implement <code>OP_CHECKCONTRACTVERIFY</code>.
+
+When evaluating <code>OP_CHECKCONTRACTVERIFY</code> (<code>OP_SUCCESS187</code>,
+<code>0xbb</code>), the expected format of the stack, shown top to bottom, is:
+
+<source>
+<flags>
+<taptree>
+<pk>
+<index>
+<data>
+</source>
+
+where:
+
+* <code><flags></code> is a minimally encoded integer, according to one of the values defined below.
+* <code><taptree></code> is the Merkle root of the taproot tree, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><pk></code> is called the ''naked key'', and it's a valid 32-byte x-only public key, or a minimally encoded <code>-1</code>, or the empty buffer.
+* <code><index></code> is a minimally encoded -1, or a minimally encoded non-negative integer.
+* <code><data></code> is a buffer of arbitrary length.
+
+In short, the semantics of the opcode with respect to the Script can be summarized as follows:
+
+<blockquote>
+Verify that the input/output with the given 'index' is a P2TR UTXO whose taproot output key is obtained from 'pk', tweaked with the hash of 'data' (if non-empty), then taptweaked with 'taptree' (if non-empty).
+</blockquote>
+
+If the <code><data></code> is non-empty, then the additive tweak for the data is computed as:
+
+<source lang="python">
+  data_tweak = sha256(pk || data)
+</source>
+
+In the following, the ''current input'' is the input whose Script is being executed.
+
+The following value of the <code><flags></code> are defined:
+* <code>CCV_FLAG_CHECK_INPUT = -1</code>: Check an input's script; no amount check.
+* <code>CCV_FLAG_CHECK_OUTPUT = 0</code>: Check an output's script; preserve the (possibly residual) amount.
+* <code>CCV_FLAG_IGNORE_OUTPUT_AMOUNT = 1</code>: Check an output's script; ignore amounts.
+* <code>CCV_FLAG_DEDUCT_OUTPUT_AMOUNT = 2</code>: Check an output's script; deduct the output amount from the input's residual amount.
```

Comment:
For these two I assume `CCV_FLAG_CHECK_OUTPUT` must be set? And they can't be both set.